book/testing_code.md

```diff
@@ -74,152 +86,209 @@ including https://www.jeremyjordan.me/testing-ml/
 [#15](https://github.com/best-practice-and-impact/qa-of-code-guidance/issues/15)
 ```
 
-## When to run tests
 
-Tests should be run whenever you make changes to your project.
-This ensures that changes do not break the existing, intended functionality of your code.
-When tests fail, you should endeavour to fix these before adding these changes to a stable or production version of your code.
+## Structure your tests to reflect how the code works
 
-If you have intentionally altered the functionality of your code, this will likely break existing tests.
-Failing tests here are a good reminder that your should update your documentation and tests to reflect the new functionality.
+Tests come in many shapes and sizes, but usually follow the pattern:
+1. Arrange - set up any objects for your test, e.g. example input data and expected output data.
+2. Act - run the code that you are testing (1 or more functions or methods).
+3. Assert - verify that the code performed the expected action, e.g. that the output matched the expected output.
 
-If your collection of tests runs quickly, it's simplest to run them all often.
-If some tests take considerably longer to run, you might want to run these less often - perhaps only when relevant changes have been made.
-Otherwise, running the entire collection of tests has the added benefit of capturing unexpected side-effects of your changes.
-For example, you might pick up an unexpected failure in part of your code that you have not directly changed.
-
-It's not easy to remember to run your tests at regular intervals.
-You're already putting effort into `commit`ing your changes to a version control system regularly.
-And you're right to think "surely this could be automated too?"
-[Continuous integration](continuous-integration) can be used to automate testing, amongst other quality assurance measures, and can be triggered when changes are made to your remote version control repository.
-These tools can be used to ensure that all changes to a particular project are tested.
-Additionally, it allows others that are reviewing your code to see the results of your tests.
-
-An alternative to continuous integration, is using a Git hook.
-[Git hooks](https://git-scm.com/docs/githooks) are scripts that can be set to run locally at specific points in your Git workflow.
-For example, we might set up a `pre-commit` or `pre-push` hook that runs our tests before we make each commit or push to the remote repository.
-This might stop our commit/push if the tests fail, so that we don't push breaking changes to our remote repository.
+We'll use a couple of examples throughout this section of the book, to demonstrate how tests relate to analytical code. 
+Here we've a basic function to get the absolute equivalent of a number:
 
+````{tabs}
```

Comment:
These functions return negative numbers when inputs are negative - is that intentional? If so, it isn't currently clear.

Reply:
you wouldn't believe I did a mathematical subject :/